### PR TITLE
Use AVA for integration tests

### PR DESCRIPTION
**Pull request by dougmoscrop**
## What did you implement:

I made the integration tests run much faster by running in parallel (up to 8 at a time).

## How can we verify it:

Run the complex integration tests.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Enable ["Allow edits from maintainers"](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) for this PR
- [x] Change ready for review message below

***Is this ready for review?:*** YES

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -49,10 +49,11 @@
   "scripts": {
     "test": "istanbul cover -x '**/*.test.js' node_modules/mocha/bin/_mocha tests/all -- -R spec --recursive",
     "lint": "eslint .",
-    "simple-integration-test": "mocha tests/integration/simple-integration-test",
-    "complex-integration-test": "mocha tests/integration/all"
+    "simple-integration-test": "ava tests/integration/simple-integration-test.js",
+    "complex-integration-test": "ava -c 10 tests/integration/general/**/tests.js tests/integration/aws/**/tests.js"
```
Comment (flomotlik) [inline package.json]: any reason for splitting general and aws tests here? If we add another folder it wouldn't be picked up in this case.

**Comment by dougmoscrop** [comment]
Nope just that the simple integration test was separate. If you're ok with
complex running ALL then van change the glob

**Comment by dougmoscrop** [comment]
I've included `scripts/integration-test-cleanup.js` which I found helpful because when the tests fail they sometimes fail to clean up after themselves and a lot of extraneous resources get left behind.

I changed the way the utils names the test services slightly - this is because starting tests in parallel caused stacks with colliding names to be created. I tried using uuids to start with, but those made various resource names too long (such as permissions). So I used a high resolution timer to get a shorter-but-hopefully-unique stack name. So far so good.

The integration test cleanup script uses a pattern that matches this naming convention, but if you want to use it to clean up older stuff, it takes the pattern as an environment variable too.

**Comment by nikgraf** [comment]
@dougmoscrop this is awesome

We discussed it in a team call yesterday. Basically we want this in, but there are two concerns:
- Could Jest be even fast with it's caching strategy?
- What about the unit-tests? We want to avoid having two different test-runners to make it easier for all contributors.

Early next week I want to see how fast the test suite is using Jest. Depending on this outcome we either merge yours or migrate the unit-tests to Ava or the integration test suite to Jest and afterwards the unit-tests too.

Do you have any thoughts on Jest vs Ava?

**Comment by dougmoscrop** [comment]
- I did the integration tests because they needed it the most and I didn't want to do a 'big bang' PR for all the unit tests too, but the goal was eventual convergence

- as for jest vs. ava, I've never used jest is all :)  the auto-mocking felt like it might be complicated for integration tests?

**Comment by nikgraf** [comment]
Makes sense :)

Jest actually doesn't automock since a couple versions :) This could help to transform the whole testsuite.

**Comment by nikgraf** [comment]
https://www.npmjs.com/package/jest-codemods

**Comment by dougmoscrop** [comment]
oh cool :)

so this was just a way to test that it was valuable to parallelize. it is! interested in seeing where jest goes.

**Comment by nikgraf** [comment]
Thanks for your awesome work here @dougmoscrop. I based a Jest PR https://github.com/serverless/serverless/pull/2975 on yours and going to close this one in favour of it.

**Pull request closed without merging**